### DISCUSSION TITLE
SSR error handling

### DISCUSSION THREAD

**Discussion by Godsenal** [TanStack/query#3448] [Q&A]
Hi!

I'm using react-query in SSR environment very happily.

I refer this [SSR document](https://react-query.tanstack.com/guides/ssr#only-successful-queries-are-included-in-dehydration) a lot.

I'm confused difference between queryClient's error state and dehydrated error state.

When error occurred in SSR (prefetchQuery or fetchQuery), queryClient's error state updated, but dehydrated state is not.
So, when component rendered in SSR, `useQuery` return error state, but when rendered in client side(CSR) error would be `undefined` before fetching.

In SSR Documents,

> Any query with an error is automatically excluded from dehydration. This means that the default behavior is to pretend these queries were never loaded on the server, usually showing a loading state instead, and retrying the queries on the queryClient. This happens regardless of error.

If this is desirable behavior, shouldn't `useQuery` return error as `undefined`  like dehydrated state ?

This also would be problem when use `useErrorBoundary` option. 

When use `useErrorBoundary` option is SSR, it would throw error in SSR (which doesn't support ErrorBoundary) because of queryClient's error state. 

This wouldn't be a problem in Next.js because queryClient will be newly created in each page's `getServerSideProps`. But in custom [SSR frameworks guide](https://react-query.tanstack.com/guides/ssr#on-the-server), it shows same QueryClient used both prefetch and render.

**Comment by TkDodo** [comment]
> When error occurred in SSR (prefetchQuery or fetchQuery), queryClient's error state updated, but dehydrated state is not.

the dehydration doesn't have any state. If you dehydrate a queryClient, you will get a json structure that's detached from the client. It will contain everything that was dehydrated as part of the dehydrateOptions. As you've seen, errors are excluded, but you can include them. See [the docs here](https://react-query.tanstack.com/reference/hydration#limitations).

> If this is desirable behavior, shouldn't useQuery return error as undefined like dehydrated state ?

errors are just not included in the dehydration, so when you hydrate again on the client and call `useQuery` on that key - there is nothing in the cache for that key - it's empty. So useQuery will go into `loading` state and fetch the data on the client side.

 > When use useErrorBoundary option is SSR, it would throw error in SSR (which doesn't support ErrorBoundary) because of queryClient's error state.

`useErrorBoundary` does nothing for `prefetchQuery` so it shouldn't be a problem.

**Comment by Godsenal** [comment]
> errors are just not included in the dehydration, so when you hydrate again on the client and call useQuery on that key - there is nothing in the cache for that key - it's empty. So useQuery will go into loading state and fetch the data on the client side.

`useQuery` is empty when render on client like you said.
But when rendered on Server(`renderToString`), it returns `error` if prefetched query has error. 

> useErrorBoundary does nothing for prefetchQuery so it shouldn't be a problem.

This could be problem when error occurred on `prefetchQuery` and  `useErrorBoundary` options set for `useQuery` which rendered on Server.

```ts
const fn = () => {
  throw new Error('error!');
}

const App = () => {
  // whice throw error on server, because QueryClient's queryState has error on this query key.
  const {error} = useQuery('key', fn, {useErrorBoundary: true});

   return <h1>{error?.message}</h1>
}


// server code from the doc

 import { dehydrate, Hydrate, QueryClient, QueryClientProvider } from 'react-query';
 
 function handleRequest (req, res) {
   const queryClient = new QueryClient()
   await queryClient.prefetchQuery('key', fn)
   const dehydratedState = dehydrate(queryClient)
 
   const html = ReactDOM.renderToString(
     <QueryClientProvider client={queryClient}>
       <Hydrate state={dehydratedState}>
         <App />
       </Hydrate>
     </QueryClientProvider>
   )
 
   res.send(`
     <html>
       <body>
         <div id="root">${html}</div>
         <script>
           window.__REACT_QUERY_STATE__ = ${JSON.stringify(dehydratedState)};
         </script>
       </body>
     </html>
   `)
 
   queryClient.clear()
 }


```

**Reply by TkDodo** [reply]
> This could be problem when error occurred on prefetchQuery and useErrorBoundary options set for useQuery which rendered on Server.

to my knowledge, `useErrorBoundary` doesn't do anything on `prefetchQuery`, as it's only used in the react bindings for `useQuery` (`useBaseQuery`).

It would be best if you could show a codesandbox reproduction of your issue.

**Comment by Godsenal** [comment]
> to my knowledge, useErrorBoundary doesn't do anything on prefetchQuery, as it's only used in the react bindings for useQuery (useBaseQuery).

Yes. What I'm saying is,

1. error occurred on `prefetchQuery` 
2. render `<App />` with `renderToString`
3. In `<App />`,  component has `useQuery` with the same key used by `prefetchQuery`

Here is reproduction codesandbox.

https://codesandbox.io/s/razzle-examples-basic-forked-ppgz7z?file=/src/App.js

It seems that `QueryClient` stores the error state of `prefetchQuery` unlike dehydratedState. So `useErrorBoundary` option causes an error to be thrown on the server side.

Also, if error used in the render (like `<h1>{error.message}</h1>`), hydarte error occurs due to a state mismatch between `QueryClient` and dehydrate state.

In Nextjs, it is not a problem because the queryClient used for rendering and the queryClient used for prefetching are different. But in custom [SSR frameworks guide](https://react-query.tanstack.com/guides/ssr#on-the-server), it shows same QueryClient used both prefetch and render.

Should I use different QueryClient for render and prefetchQuery?

**Reply by TkDodo** [reply]
> Should I use different QueryClient for render and prefetchQuery?

Yes, I think that's it.
